### PR TITLE
Vchad experiments

### PR DESCRIPTION
**Pull request by Derposoft**
the ultimate final form, as far as i am concerned tbh

### CODE REVIEW

**Comment by Derposoft** [comment]
+argparse because i'm cool
+changed the default args to be the best hyperparameters i could find (so far)
+added k-folds
+made it so that it chooses the metrics from the epoch where it got the highest f1 score for its calculation where it averages the metrics for the final scores

**Comment by Derposoft** [comment]
also,
+layer norm support for convlstm at both cell level and lstm output level (but they are bad)

File: models/conv_lstm.py [hunk 1]
```diff
@@ -43,11 +52,21 @@ def forward(self, x):
...
+            #y = torch.permute(y, [1, 0, 2])
+        else:
+            y, (hn, cn) = self.lstm(y) # defaulting to h_0, c_0 = 0, 0
+        if self.out_norm:
+            y = self.ln(y)
 
         # decode
-        return self.o_proj(y)
+        y = self.o_proj(y)
+        return self.sigmoid(y).reshape([y.shape[0], y.shape[1]])
```
Comment (qwewqa) [inline models/conv_lstm.py]: Could use `squeeze(2)` instead of `reshape()`.

File: train.py [hunk 2]
```diff
@@ -1,90 +1,167 @@
...
-TRAIN_RATIO = 0.5 # [0.0, 1.0] proportion of data used for train
-ADD_EXTRA_ONES = 30 # labels in [anomaly_time, anomaly_time+ADD_EXTRA_ONES] are changed to 1
-'''
-SAVE=True caches a copy of the data for each time you run it with a different ADD_EXTRA_ONES setting.
-makes for faster experiments and lower data loading times, but it's not too much faster and each
-copy of the data is around 400mb. so up to you.
-'''
-SAVE = False
+###   cli arguments   ###
+args = argparse.ArgumentParser()
```
Comment (qwewqa) [inline train.py]: Why isn't this under `if __name__ == '__main__'`?

File: train.py [hunk 3]
```diff
@@ -1,90 +1,167 @@
...
+
+def create_conv_model():
+    return ConvLSTM(n_feats, KERNEL_SIZE, EMBEDDING_SIZE, N_LAYERS, 
+        cell_norm=CELL_NORM, out_norm=OUT_NORM).to(device)
 
 if __name__ == '__main__':
-    train_loader, test_loader = read_data(
+    #train_loader, test_loader = 
+    data = get_data(
         './data/features_15S.csv.gz',
```
Comment (qwewqa) [inline train.py]: You could add data path to argparse

File: models/conv_lstm.py [hunk 4]
```diff
@@ -14,6 +16,9 @@ def __init__(self, num_feats, conv_kernel_size, embedding_size, num_layers):
         self.conv_kernel_size = conv_kernel_size
         self.embedding_size = embedding_size
         self.num_layers = num_layers
+        self.dropout = dropout
```
Comment (qwewqa) [inline models/conv_lstm.py:19]: Where is this being used?

File: models/conv_lstm.py [hunk 5]
```diff
@@ -25,12 +30,16 @@ def __init__(self, num_feats, conv_kernel_size, embedding_size, num_layers):
         self.pool = nn.MaxPool1d(2, 1)
 
         # detecting
-        self.lstm = nn.LSTM(embedding_size, embedding_size, num_layers, batch_first=True)
+        if cell_norm:
```
Comment (qwewqa) [inline models/conv_lstm.py:33]: `dropout` could be passed to lstm, granted it doesn't do anything if it's only one layer since it's only applies to layers other than the final layer.

File: train.py [hunk 6]
```diff
@@ -1,90 +1,167 @@
...
+config = args.parse_args()
+
+
+###   hyperparameters   ###
+# conv model
+EMBEDDING_SIZE = config.embedding_size
+N_LAYERS = config.n_layers
+N_EPOCHS = config.n_epochs
+KERNEL_SIZE = config.kernel_size
+DROPOUT = config.dropout
```
Comment (qwewqa) [inline train.py]: Doesn't seem to be used

**Review by qwewqa — COMMENTED** [review]
Looks good overall, except dropout seems to never actually get used.

File: models/conv_lstm.py (hunk 5 above)
Comment (Derposoft) [reply models/conv_lstm.py:33]: +1, i forgot to pass it in. i'll keep it in case we end up using n_layers=2 in the future, although it still sucks that it won't be usable for the layernorm lstm since it hasn't been implemented there. it'll probably be ok since use of layernorm doesn't help anyways.

**Review by qwewqa — APPROVED** [review]
